networkx/linalg/algebraicconnectivity.py

```diff
@@ -244,7 +244,7 @@ def P(x):
             W -= (W.T * X * X.T).T
             project(W)
             # Compute the diagonal of P * L * P as a Jacobi preconditioner.
-            D = L.diagonal()
+            D = L.diagonal().astype(float)
             D += 2. * (asarray(X) * asarray(W)).sum(axis=1)
```

Comment:
It doesn't look like efficiency.  Perhaps that's the case for the W equations a couple lines up.
In fact....  it looks like ```D``` is never used. The comments say it is being computed as a preconditioner, but it never gets used. Perhaps it is a left-over from a trial-phase or debugging or timing?  

I agree with @chebee7i that if L needs to be a float that it be converted early ( like when it is created -- maybe we need a feature for that). I prefer to let python figure out which type is best when it doesn't matter -- explicit conversions in the middle of a long routine make it hard for me to keep track and may lead to efficiency issues if we end up converting back and forth (which this case don't do). 

Can someone verify that D is just not used?  If so, let's remove these lines or at least comment them out.